File: php/bob/bob.php
<?php

class Bob
{
    function respondTo(string $input): string
    {
        $trimed = trim($input);

        if (empty($trimed)) {
            return "Fine. Be that way!";
        }

        $alpha = self::isAlpha($trimed);
        $shouting = self::isShouting($trimed);
        $question = self::isQuestion($trimed);

        if ($alpha && $shouting && $question) {
            return "Calm down, I know what I'm doing!";
        }

        if ($alpha && $shouting) {
            return "Whoa, chill out!";
        }

        if ($question) {
            return "Sure.";
        }

        return "Whatever.";
    }

    static function isAlpha(string $input): bool
    {
        return preg_match("/[a-zA-Z]/", $input) == 1;
    }

    static function isShouting(string $input): bool
    {
        return $input === strtoupper($input);
    }

    static function isQuestion(string $input): bool
    {
        return substr($input, -1) === "?";
    }
}

File: erlang/bob/src/bob.erl
-module(bob).

-export([response/1]).

response(Sentence) ->
    Trimed = string:trim(Sentence),
    case {is_shouting(Trimed), is_question(Trimed), Trimed}
        of
        {_, _, []} -> "Fine. Be that way!";
        {true, true, _} -> "Calm down, I know what I'm doing!";
        {true, false, _} -> "Whoa, chill out!";
        {false, true, _} -> "Sure.";
        {_, _, _} -> "Whatever."
    end.

is_question([]) -> false;
is_question(Sentence) -> lists:last(Sentence) =:= $?.

is_shouting([]) -> false;
is_shouting(Sentence) ->
    Letters = re:replace(Sentence,
                         "[^A-Za-z]",
                         "",
                         [global, {return, list}]),
    case Letters of
        [] -> false;
        _ -> Letters == string:to_upper(Letters)
    end.
